lib/puppet/util/windows/process.rb

```diff
@@ -224,6 +224,58 @@ def windows_major_version
   end
   module_function :windows_major_version
 
+  ENVSTRINGS_TERMINATOR_WCHAR = [0,0]
+  # Returns a hash of the current environment variables encoded as UTF-8
+  # The memory block returned from GetEnvironmentStringsW is double-null terminated and the vars are paired as below;
+  # Var1=Value1\0
+  # Var2=Value2\0
+  # VarX=ValueX\0\0
+  # Note - Some env variable names start with '=' and are safe to ignore
+  # Note - The env_ptr MUST be freed using the FreeEnvironmentStringsW function
+  # Note - There is no technical limitation to the size of the environment block returned.
+  #   However a pracitcal limit of 64K is used as no single enviornment variable can exceed 32KB
+  def get_environment_strings
+    env_ptr = GetEnvironmentStringsW()
+
+    env_hash = {}
+    env_ptr.read_arbitrary_wide_string_up_to(65534, ENVSTRINGS_TERMINATOR_WCHAR).split(?\x00).each do |env_pair|
+      pair = env_pair.split("=")
+      unless pair[0] == ""
```

Comment:
Can shorten this up to:

```ruby
pairs = env_ptr.read_arbitrary_wide_string_up_to(65534, ENVSTRINGS_TERMINATOR_WCHAR)
  .split(?\x00)
  .reject { |env_str| env_str.nil? || env_str[0] == '=' }
  .map { |env_pair| env_pair.split('=', -2) }
Hash[ pairs ]
```

The `-2` in split ensures there's always 2 members in the array after the split, and if it has to fill in a second element, its an empty string.

And of course `Hash[ pairs]` is a little Ruby trick that can initialize a hash with an array of 2 element arrays.